autolab/Autolab
# Group setting update API and controller endpoint guards

## Description
This PR adds the necessary elements to allow/disallow students to change group assignment.
This PR also adds guards against students attempting to change group through calling API directly when student group self-assignment is disabled.

## Motivation and Context
To fully enable the feature of disallowing students to assign groups for assessments.

## How Has This Been Tested?
Note that this PR does not include frontend change. The test below makes use of this fact and expects failure to make changes to groups.
1. Run `rails db:migrate`
2. Run `rails c` and disallow students to change groups for an assessment of your liking by using ORM.
3. Alternatively, call the assessments endpoint `update_assessment_group_setting` with the appropriate payload parameter `allow_student_assign_group`.
4. Act as any student and attempt to create groups. Failure and a flash error message is expected.
5. Act as instructor and create new groups. (Currently, the flow is a bit messy. The expectation is that, an instructor creates a group including themself, instructor leaves the group, and invite another student to the group.) Note that when students are disallowed to assign groups, instructor DOES NOT wait for student's confirmation of membership -- they are automatically added to the group.
6. Act as a student without a group yet and attempt to join other groups. Failure and a flash error message are expected.
7. Act as a student with a group and attempt to leave their current group. Failure and a flash error message are expected.
8. Act as a student with a group and update group name. This SHOULD succeed.

## Types of changes


## Checklist:


## Other issues / help required




If unsure, feel free to submit first and we'll help you along.
